File: api/mounts.go
package vaku

import (
	"errors"
	"strconv"
	"strings"
)

var (
	// ErrMountInfo when failing to get mount information about a path.
	ErrMountInfo = errors.New("mount info")
	// ErrListMounts when failing to list vault mounts.
	ErrListMounts = errors.New("list mounts")
	// ErrNoMount when path cannot be matched to a mount.
	ErrNoMount = errors.New("no matching mount")
	// ErrRewritePath when failing to rewrite the path with mount data.
	ErrRewritePath = errors.New("rewriting path")
	// ErrMountVersion when an operation is not supported on the mount version.
	ErrMountVersion = errors.New("mount version does not support operation")
)

// mountVersion represents possible vault kv mount versions.
type mountVersion int

const (
	mv0 mountVersion = iota
	mv1
	mv2
)

// vaultOperation represents a call made to vault (read, write, delete, etc...).
type vaultOperation int

// operations that are used in vaku.
const (
	vaultList vaultOperation = iota + 1
	vaultRead
	vaultWrite
	vaultDelete
	vaultDestroy
	vaultDeleteMeta
	vaultReadMeta
)

// mountInfo takes a path and returns the mount path and version.
func (c *Client) mountInfo(p string) (string, mountVersion, error) {
	mounts, err := c.mountProvider.ListMounts()
	if err != nil {
		return "", mv0, newWrapErr(p, ErrMountInfo, newWrapErr(err.Error(), ErrListMounts, nil))
	}

	for _, mount := range mounts {
		// Ensure '/' so that no match on foo/bar/ when actual path is foo/barbar/
		mount.Path = EnsureFolder(mount.Path)
		if strings.HasPrefix(p, mount.Path) {
			if mount.Version == "" {
				return mount.Path, mv0, nil
			}

			return mount.Path, mountStringToVersion(mount.Version), nil
		}
	}

	return "", mv0, newWrapErr(p, ErrMountInfo, newWrapErr(p, ErrNoMount, nil))
}

// mountStringToVersion converts a mount version string from vault to a MountVersion.
func mountStringToVersion(v string) mountVersion {
	version, err := strconv.Atoi(v)
	if err != nil {
		return mv0
	}

	return mountVersion(version)
}

// rewritePath rewrites a vault input path based on the mount version and operation.
func (c *Client) rewritePath(p string, op vaultOperation) (string, mountVersion, error) {
	mount, version, err := c.mountInfo(p)
	if err != nil {
		return "", mv0, newWrapErr(p, ErrRewritePath, err)
	}

	// check if the operation is supported
	if !mountSupportsOperation(op, version) {
		return "", version, newWrapErr(p, ErrMountVersion, nil)
	}

	// only rewrite mv2 mounts
	if version != mv2 {
		return p, version, nil
	}

	switch op {
	case vaultList, vaultDeleteMeta, vaultReadMeta:
		p = InsertIntoPath(p, mount, "metadata")
	case vaultRead, vaultWrite, vaultDelete:
		p = InsertIntoPath(p, mount, "data")
	case vaultDestroy:
		p = InsertIntoPath(p, mount, "destroy")
	}

	return p, version, nil
}

func mountSupportsOperation(op vaultOperation, v mountVersion) bool {
	// v2 mounts support all operations
	if v == mv2 {
		return true
	}

	// v1 mounts don't support these
	switch op {
	case vaultDestroy, vaultDeleteMeta, vaultReadMeta:
		return false
	}

	// default to supported
	return true
}

File: src/api/mounts.rs
//! Mount-info lookup and KV-v2 path rewriting.

use crate::api::error::{Error, ErrorKind};
use crate::api::helpers::{ensure_folder, insert_into_path};
use crate::api::secret::{KV2_DATA, KV2_DESTROY, KV2_METADATA};

/// KV mount versions recognized by Vaku.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MountVersion {
    Mv0,
    Mv1,
    Mv2,
    Other(i64),
}

impl MountVersion {
    pub fn from_str(s: &str) -> Self {
        match s.parse::<i64>() {
            Ok(0) => MountVersion::Mv0,
            Ok(1) => MountVersion::Mv1,
            Ok(2) => MountVersion::Mv2,
            Ok(n) => MountVersion::Other(n),
            Err(_) => MountVersion::Mv0,
        }
    }
}

/// Vault operations supported by Vaku. Used to drive path rewriting.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum VaultOp {
    List,
    Read,
    Write,
    Delete,
    Destroy,
    DeleteMeta,
    ReadMeta,
}

/// Returns true when the operation is supported on the given mount version.
pub fn mount_supports_operation(op: VaultOp, v: MountVersion) -> bool {
    if matches!(v, MountVersion::Mv2) {
        return true;
    }
    !matches!(op, VaultOp::Destroy | VaultOp::DeleteMeta | VaultOp::ReadMeta)
}

/// Find the mount that `path` lives under. Returns `(mount_path, version)`.
pub async fn mount_info(
    provider: &dyn crate::api::mount_provider::MountProvider,
    path: &str,
) -> Result<(String, MountVersion), Error> {
    let mounts = provider.list_mounts().await.map_err(|e| {
        Error::wrap(
            path,
            ErrorKind::MountInfo,
            Some(Box::new(Error::wrap(
                &e.to_string(),
                ErrorKind::ListMounts,
                None,
            ))),
        )
    })?;

    for mount in mounts {
        // Always check against the folder form so we don't match a prefix like
        // `foo/bar/` against the path `foo/barbar/...`.
        let mp = ensure_folder(&mount.path);
        if path.starts_with(&mp) {
            let version = if mount.version.is_empty() {
                MountVersion::Mv0
            } else {
                MountVersion::from_str(&mount.version)
            };
            return Ok((mp, version));
        }
    }

    Err(Error::wrap(
        path,
        ErrorKind::MountInfo,
        Some(Box::new(Error::wrap(path, ErrorKind::NoMount, None))),
    ))
}

/// Rewrite `p` for the given operation, transparently inserting KV v2's
/// `data` / `metadata` / `destroy` sub-paths when required.
pub async fn rewrite_path(
    provider: &dyn crate::api::mount_provider::MountProvider,
    p: &str,
    op: VaultOp,
) -> Result<(String, MountVersion), Error> {
    let (mount, version) = match mount_info(provider, p).await {
        Ok(v) => v,
        Err(e) => return Err(Error::wrap(p, ErrorKind::RewritePath, Some(Box::new(e)))),
    };

    if !mount_supports_operation(op, version) {
        return Err(Error::wrap(p, ErrorKind::MountVersion, None));
    }

    if !matches!(version, MountVersion::Mv2) {
        return Ok((p.to_string(), version));
    }

    let rewritten = match op {
        VaultOp::List | VaultOp::DeleteMeta | VaultOp::ReadMeta => {
            insert_into_path(p, &mount, KV2_METADATA)
        }
        VaultOp::Read | VaultOp::Write | VaultOp::Delete => insert_into_path(p, &mount, KV2_DATA),
        VaultOp::Destroy => insert_into_path(p, &mount, KV2_DESTROY),
    };

    Ok((rewritten, version))
}
